soccer/planning/CollectPathPlanner.cpp

```diff
@@ -0,0 +1,295 @@
+        averageBallVelInitialized = true;
+    } else {
+        averageBallVel = *_targetPointAveragingGain * averageBallVel + 
+                         (1 - *_targetPointAveragingGain) * ball.vel;
+    }
+
+    // Do the transitions
+    float dist = (startInstant.pos - ball.pos).mag() - Robot_MouthRadius - *_approachDistTarget;
+    float speedDiff = (startInstant.vel - averageBallVel).mag() - *_touchDeltaSpeed;
+
+    // If we are close enough to the target point near the ball
+    // and almost the same speed we want, start slowing down
+    // TODO: Check for ball sense?
+    if (dist < *_distCutoffToControl && speedDiff < *_velCutoffToControl && currentState == Approach) {
+        currentState = Control;
+    }
+
+    // Check if we need to go back into approach
+    if ((ball.pos - startInstant.pos).mag() > 2*Robot_Radius && currentState == Control) {
+        cout << "Reseting" << endl;
+        currentState = Approach;
+    }
+
+    // Approach direction is the direction we move towards the ball and through it
+    // Only set once
+    // Reset when the ball state changes significantly (This is done in python)
+
+    // If we haven't found an approach direction yet
+    if (approachDirection == Point(0,0)) {
+        if (ball.vel.mag() < *_ballSpeedApproachDirectionCutoff) {
+            // Move directly to the ball
+            approachDirection = (ball.pos - startInstant.pos).norm();
+        } else {
+            // Approach the ball from behind
+            // TODO: May need to rethink average here since it'll be based on the first ball vel only
+            approachDirection = averageBallVel.norm();
+        }
+    }
+
+    switch (currentState) {
+    // Moves from the current location to the ball
+    // Controls the robot up until just before the point of contact
+    case Approach: {
+
+        // The target position shouldn't be the ball, it should be where the mouth
+        // is touching the ball
+        // Move to ball position matching ball speed
+
+        // Setup targets for path planner
+        Point targetPos = ball.pos - (*_approachDistTarget + Robot_MouthRadius) * approachDirection;
+        MotionInstant target(targetPos, averageBallVel + approachDirection.norm() * *_touchDeltaSpeed);
+        vector<Point> startEndPoints{startInstant.pos, target.pos};
+
+        // Decrease accel over the entire path
+        // In the correct case, only the end will be decreased
+        motionConstraints.maxAcceleration *= *_approachAccelScalePercent;
+
+        // Build a path from now to the end
+        unique_ptr<Path> path =
+            RRTPlanner::generatePath(startEndPoints, obstacles, motionConstraints, startInstant.vel, target.vel);
+
+        path->setDebugText("approaching");
+
+        // Append to the partial path
+        if (path) {
+            if (partialPath) {
+                path = make_unique<CompositePath>(move(partialPath),
+                                                  move(path));
```

Comment:
Is this expected to run every frame? If so this is going to turn the path into basically a linked-list, right?

Reply:
@JNeiger I can't remember, did we talk about this in person?